cluster/reconcile.go

```diff
@@ -0,0 +1,134 @@
+package cluster
+
+import (
+	"fmt"
+
+	"github.com/rancher/rke/hosts"
+	"github.com/rancher/rke/k8s"
+	"github.com/rancher/rke/services"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh"
+	"k8s.io/client-go/kubernetes"
+)
+
+func ReconcileCluster(kubeCluster, currentCluster *Cluster) error {
+	logrus.Infof("[reconcile] Reconciling cluster state")
+	if currentCluster == nil {
+		logrus.Infof("[reconcile] This is newly generated cluster")
+
+		return nil
+	}
+	// to handle if current local admin is down and we need to use new cp from the list
+	if !isLocalConfigWorking(kubeCluster.LocalKubeConfigPath) {
+		if err := rebuildLocalAdminConfig(kubeCluster); err != nil {
+			return err
+		}
+	}
+
+	kubeClient, err := k8s.NewClient(kubeCluster.LocalKubeConfigPath)
+	if err != nil {
+		return fmt.Errorf("Failed to initialize new kubernetes client: %v", err)
+	}
+
+	key, _ := checkEncryptedKey(kubeCluster.SSHKeyPath)
+
+	if err := reconcileWorker(currentCluster, kubeCluster, key, kubeClient); err != nil {
```

Comment:
Any particular reason workers are being reconciled first? The official upgrade doc advises to upgrade etcd/master first: https://kubernetes.io/docs/getting-started-guides/ubuntu/upgrades/

Reply:
This is for the cases where a host has both controller and worker role, and this same host is the only controller host, when the user run ./rke up it should first remove worker components including kubelet and kubeproxy, if we remove the controller components first it won't be able to remove the worker

Reply:
Also upgrade is not a part of reconcile, its a part of adding components